### PR TITLE
Add Shuffle and sharding datapipes to datasets

### PR DESCRIPTION
**Pull request by parmeet**
Reference Issue: #1727

### CODE REVIEW

**Comment by ejguan** [comment]
I enabled AWS extension last night. It seems the C-Extension is using glibc-2.29 during compilation. Will do a quick fix.

**Comment by parmeet** [comment]
> Perhaps it might make sense to add a small test that makes sure all datapipes come with a sharding filter and a shuffler? 

Thanks @NicolasHug  for the suggestion. I think it is indeed a good idea. Let me add it as well.

**Review by VirgileHlav — COMMENTED** [review]
Do we expect to have variations for shuffle and sharding between datasets, as for now all datasets use `.shuffle().set_shuffle(False).sharding_filter()`? If not, we could consider using instead a decorator to wrap this call on the returned object and clean up the code.

**Comment by NicolasHug** [comment]
Good point @VirgileHlav .

In general you want to shard and shuffle on light objects (before decoding, before transforms) to avoid unnecessary computations, and to save memory. For now torchtext datasets yield light objects (simple text), but maybe in the future this will change?

In torchvision, we wrap in different places for each dataset, so using a decorator isn't an option.

**Review by VirgileHlav — APPROVED** [review]
@NicolasHug thanks for the clarification, in that case let's keep it as is for now. @parmeet I will also add this to the dataset effort in #1710.

Otherwise LGTM

**Comment by parmeet** [comment]
Thanks @VirgileHlav SGTM! As @NicolasHug mentioned it could potentially vary from dataset to dataset. So if in some cases processing is needed at sample level, you could potentially shard the pipe before processing, but otherwise adding it at the end is fine.

File: test/datasets/common.py [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+        if type(dp) == tuple:
+            dp = list(dp)
+        else:
+            dp = [dp]
+
+        for dp_split in dp:
+            dp_graph = get_all_graph_pipes(traverse(dp_split))
+            for annotation_dp_type in [Shuffler, ShardingFilter]:
+                if not any(isinstance(dp, annotation_dp_type) for dp in dp_graph):
+                    raise AssertionError(f"The dataset doesn't contain a {annotation_dp_type.__name__}() datapipe.")
```
[the remark this answers is not in the document]
Comment (ejguan) [reply test/datasets/common.py:25]: Thanks. I will do a fix for manylinux1 wheel first then add linter for you.

File: test/datasets/common.py [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+from torchdata.datapipes.iter import Shuffler, ShardingFilter
+from torchtext.datasets import DATASETS
+
+from ..common.torchtext_test_case import TorchtextTestCase
+
+
+class TestShuffleShardDatasetWrapper(TorchtextTestCase):
+    # Note that for order i.e shuffle before sharding, TorchData will provide linter warning
+    # Modify this test when linter warning is available
+    @parameterized.expand(list(DATASETS.items()))
```
Comment (Nayef211) [inline test/datasets/common.py:13]: QQ: it looks like we're not using the `dataset_name` anywhere. Why don't we just pass in the `dataset_fn` to the test by doing something like

```python
@parameterized.expand(list(DATASETS.values()))
```

Comment (parmeet) [reply test/datasets/common.py:13]: Ahh my bad, thanks for the catch. Will fix it!

Comment (Nayef211) [reply test/datasets/common.py:13]: I actually realized that the `parameterized.expand` decorator complains when passing in the list of  `dataset_fn`. Lmk if you're able to figure out how to resolve the error

Comment (parmeet) [reply test/datasets/common.py:13]: Yupp, the problem is we need to pass tuples inside list. Just created PR to fix it https://github.com/pytorch/text/pull/1733

Comment (Nayef211) [reply test/datasets/common.py:13]: Awesome. Just incorporated this in my PR https://github.com/pytorch/text/pull/1732

**Pull request merged**